### PR TITLE
Bulk delete users

### PR DESCRIPTION
**Pull request by orangejenny**
##### SUMMARY
https://dimagi-dev.atlassian.net/browse/ICDS-1044

##### FEATURE FLAG
New flag: Allow bulk deletion of users based on a username upload.

@AndreaMKing5 The flag is marked solutions limited, let me know if you would prefer something else.

##### PRODUCT DESCRIPTION
Limited to users without form submissions. 
Screen Shot 2020-02-03 at 3 51 55 PM

### CODE REVIEW

File: corehq/apps/users/views/mobile/users.py [hunk 1]
```diff
@@ -1126,6 +1127,60 @@ def get(self, request, domain, *args, **kwargs):
...
+        deleted_count = 0
+        usernames_with_forms = []
+        for row in sheet:
+            try:
+                username = row['username']
+            except KeyError:
+                messages.error(request, _("Please upload a file with a 'username' column."))
+                return self.get(request, *args, **kwargs)
+
+            user = CommCareUser.get_by_username(f"{username}@{request.domain}.commcarehq.org")
```
Comment (snopoke) [inline corehq/apps/users/views/mobile/users.py]: could do a single couch request to get all users with given usernames

File: corehq/apps/users/views/mobile/users.py [hunk 2]
```diff
@@ -1126,6 +1127,60 @@ def get(self, request, domain, *args, **kwargs):
...
+            if user:
+                if FormES().domain(request.domain).user_id(user.get_id).count():
```
Comment (snopoke) [inline corehq/apps/users/views/mobile/users.py]: could do a single request to check if users have forms by doing a terms aggregation on `username` or `user_id`:

```python
user_ids_with_forms = FormES().domain(request.domain).user_id(user_ids).terms_aggregation('user_id', 'user_id').run().aggregations.user_id..keys
```

**Review by snopoke — APPROVED** [review]
Couple of general comments:

1. The validation could probably reuse the same classes as the user importer: `UsernameTypeValidator`, variation of `ExistingUserValidator`.

2. What's the intended scale of this? Will we need to update it later to be asynchronous and do the deletions in bulk?

File: corehq/apps/users/views/mobile/users.py (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: It doesn't seem to work the same, tried with a local user who has form submissions but the aggregate query doesn't return anything:
```
In [34]: FormES().domain(domain).user_id(user_id).count()
Out[34]: 4

In [35]: FormES().domain(domain).user_id([user_id]).terms_aggregation('user_id', 'user_id').run().aggregations.user_id.keys
Out[35]: []
```

File: corehq/apps/users/views/mobile/users.py (hunk 1 above)
Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: 413459bd50ea2d7f6fb0b706a6241ff7d6a05529

File: corehq/apps/users/views/mobile/users.py (hunk 2 above)
Comment (esoergel) [reply corehq/apps/users/views/mobile/users.py]: I think it needs to use [`form.meta.userID`](https://github.com/dimagi/commcare-hq/blob/3cc508d2668776578a17465fb7e8f5bf60226b87/corehq/apps/es/forms.py#L73-L79) as the user ID path.
```python
FormES().domain(domain).user_id([user_id]).terms_aggregation('form.meta.userID', 'user_id').run().aggregations.user_id.keys
```

**Comment by orangejenny** [comment]
@snopoke 

I took a stab at using the user importer validation, but the code turned out longer & more awkward.

The biggest request thus far has been ~5000 users. Does that seem small enough to do synchronously (and to hold all user docs in memory)?

File: corehq/apps/users/views/mobile/users.py (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: That fixes it, thank you!

Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: 77f2d7d7be5

Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: 77f2d7d7be5

File: corehq/apps/users/templates/users/edit_commcare_user.html [hunk 3]
```diff
@@ -178,6 +178,9 @@ <h4 class="modal-title">{% blocktrans with couch_user.human_friendly_name as fri
               <p class="alert alert-warning">
                 <i class="fa fa-exclamation-triangle"></i>
                 {% trans "Bad things will happen if you don't read this" %}
+                {% blocktrans %}
+                  It is important you read the entire message below thoroughly before completing this action.
+                {% endblocktrans %}
```
Comment (mkangia) [inline corehq/apps/users/templates/users/edit_commcare_user.html]: i thought you were replacing the "Bad things.." message with this.

File: corehq/apps/users/templates/users/mobile_workers.html [hunk 4]
```diff
@@ -56,9 +56,16 @@
             {% endif %}
             {% if can_add_extra_users and can_access_all_locations %}
               <a class="btn btn-default" href="{% url "user_fields_view" domain %}" id="btn-edit_user_fields">
+                <i class="fa fa-edit"></i>
                 {% trans "Edit User Fields" %}
               </a>
             {% endif %}
+            {% if request|toggle_enabled:"BULK_USER_DELETE" %}
```
Comment (mkangia) [inline corehq/apps/users/templates/users/mobile_workers.html]: nit: how about giving this option only to project admin to avoid another toggle and to avoid giving this big power with just a toggle add?

File: corehq/apps/users/views/mobile/users.py [hunk 5]
```diff
@@ -1126,6 +1128,71 @@ def get(self, request, domain, *args, **kwargs):
...
+            .domain(request.domain)
+            .user_id([doc['_id'] for doc in user_docs_by_id.values()])
+            .terms_aggregation('form.meta.userID', 'user_id')
+        ).run().aggregations.user_id.keys
+
+        deleted_count = 0
+        for user_id, doc in user_docs_by_id.items():
+            if user_id not in user_ids_with_forms:
+                CommCareUser.wrap(doc).delete()
+                deleted_count += 1
```
Comment (mkangia) [inline corehq/apps/users/views/mobile/users.py]: i think we should keep the list of usernames that were deleted.
And in case the request fails midway due to say, `ResourceConflict` we can at least inform the usernames for the user that were successfully deleted.

File: corehq/apps/users/views/mobile/users.py [hunk 6]
```diff
@@ -1126,6 +1128,71 @@ def get(self, request, domain, *args, **kwargs):
...
+
+    @property
+    def page_context(self):
+        context = self.main_context
+        context.update({
+            'bulk_upload_form': get_bulk_upload_form(),
+        })
+        return context
+
+    def post(self, request, *args, **kwargs):
```
Comment (mkangia) [inline corehq/apps/users/views/mobile/users.py]: there is a lot happening happening in this method which makes it a little hard to follow and possibly brittle

How about either
1. breaking it into methods
2. making a utility class that handles this request.

I would vote for utility class, @esoergel i know would vote for methods because he likes methods :)

File: corehq/apps/users/views/mobile/users.py [hunk 7]
```diff
@@ -1126,6 +1128,71 @@ def get(self, request, domain, *args, **kwargs):
...
+            sheet = workbook.get_worksheet()
+        except WorksheetNotFound:
+            messages.error(request, _("Workbook has no worksheets"))
+            return self.get(request, *args, **kwargs)
+
+        try:
+            usernames = [format_username(row['username'], request.domain) for row in sheet]
+        except KeyError:
+            messages.error(request, _("No users found. Please check your file contains a 'username' column."))
+            return self.get(request, *args, **kwargs)
```
Comment (mkangia) [inline corehq/apps/users/views/mobile/users.py]: Can't the above three validations go in the form instead?

**Comment by mkangia** [comment]
wohoo!

thanks for shipping this @orangejenny 
I am not entirely sure about the speed with which this would handle deleting the users. I have done few deletes and they are pretty quick usually. Having it async would be a good proactive step, but i think we can just let this be sync for now and move to async if needed.

File: corehq/apps/users/views/mobile/users.py [hunk 8]
```diff
@@ -1126,6 +1128,71 @@ def get(self, request, domain, *args, **kwargs):
...
+
+        user_docs_by_id = {doc['_id']: doc for doc in get_user_docs_by_username(usernames)}
+        usernames_not_found = set(usernames).difference(set([doc['username'] for doc in user_docs_by_id.values()]))
+        user_ids_with_forms = (
+            FormES()
+            .domain(request.domain)
+            .user_id([doc['_id'] for doc in user_docs_by_id.values()])
```
Comment (snopoke) [inline corehq/apps/users/views/mobile/users.py]: nit: just use dict keys
```suggestion
            .user_id(list(user_docs_by_id))
```

File: corehq/apps/users/views/mobile/users.py [hunk 9]
```diff
@@ -1126,6 +1128,71 @@ def get(self, request, domain, *args, **kwargs):
...
+            return self.get(request, *args, **kwargs)
+
+        try:
+            usernames = [format_username(row['username'], request.domain) for row in sheet]
+        except KeyError:
+            messages.error(request, _("No users found. Please check your file contains a 'username' column."))
+            return self.get(request, *args, **kwargs)
+
+        user_docs_by_id = {doc['_id']: doc for doc in get_user_docs_by_username(usernames)}
+        usernames_not_found = set(usernames).difference(set([doc['username'] for doc in user_docs_by_id.values()]))
```
Comment (snopoke) [inline corehq/apps/users/views/mobile/users.py]: nit: use set comprehension
```suggestion
        usernames_not_found = set(usernames) - {doc['username'] for doc in user_docs_by_id.values()}
```

if you make `usernames` a set instead of a list you can also drop conversion of it to a set.

File: corehq/apps/users/views/mobile/users.py (hunk 7 above)
Comment (snopoke) [reply corehq/apps/users/views/mobile/users.py]: I don't think this level of validation belongs in the form

File: corehq/apps/users/views/mobile/users.py [hunk 10]
```diff
@@ -1126,6 +1128,71 @@ def get(self, request, domain, *args, **kwargs):
...
+            FormES()
+            .domain(request.domain)
+            .user_id([doc['_id'] for doc in user_docs_by_id.values()])
+            .terms_aggregation('form.meta.userID', 'user_id')
+        ).run().aggregations.user_id.keys
+
+        deleted_count = 0
+        for user_id, doc in user_docs_by_id.items():
+            if user_id not in user_ids_with_forms:
+                CommCareUser.wrap(doc).delete()
```
Comment (snopoke) [inline corehq/apps/users/views/mobile/users.py]: there's no need to wrap this:
```suggestion
                CommCareUser.get_db().delete_doc(doc)
```

You could also do this in bulk which is easy using the `IterDB`
```python
with IterDB(db) as iter_db:
    for user_id, doc in user_docs_by_id.items():
        if user_id not in user_ids_with_forms:
            iter_db.delete(doc)

error_ids = iter_db.error_ids
```

**Review by snopoke — COMMENTED** [review]
Thanks for the changes. I've made a few more comments but mostly small stuff.

File: corehq/apps/users/views/mobile/users.py (hunk 10 above)
Comment (mkangia) [reply corehq/apps/users/views/mobile/users.py]: Thats nice, handy iteration tool.

File: corehq/apps/users/views/mobile/users.py (hunk 7 above)
Comment (mkangia) [reply corehq/apps/users/views/mobile/users.py]: why so @snopoke ?
I was thinking
1. nice to have all validations at one place
2. the uploaded file is an attribute of the form so it should be validated like clean_field.

File: corehq/apps/users/templates/users/edit_commcare_user.html (hunk 3 above)
Comment (orangejenny) [reply corehq/apps/users/templates/users/edit_commcare_user.html]: oops - a1aface3255

File: corehq/apps/users/templates/users/mobile_workers.html (hunk 4 above)
Comment (orangejenny) [reply corehq/apps/users/templates/users/mobile_workers.html]: I don't like adding a new flag either, but project admin isn't usually the right place to add things, different projects will have different ideas of what a project admin should do.

File: corehq/apps/users/views/mobile/users.py (hunk 5 above)
Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: I think this is okay as it is; I'd rather not list out all the success where there are hundreds of them. If there's a failure partway through the user can re-upload the same file.

File: corehq/apps/users/views/mobile/users.py (hunk 9 above)
Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: e2310e4fa97

Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: e2310e4fa97

File: corehq/apps/users/views/mobile/users.py (hunk 6 above)
Comment (esoergel) [reply corehq/apps/users/views/mobile/users.py]: I like _functions_ :grin: (or staticmethods, which are just namespaced functions).  Methods can be nice too, when not complicated by inheritance or mutating instance variables.  Methods are sometimes necessary where there are a lot of flag arguments or memoization.

File: corehq/apps/users/views/mobile/users.py (hunk 10 above)
Comment (esoergel) [reply corehq/apps/users/views/mobile/users.py]: Though if not wrapped, the extra stuff that happens when `delete()` is called will need to be done manually.  Notably, it looks like that's how the django user is deleted:
https://github.com/dimagi/commcare-hq/blob/5361705c913595e5dc6e0498560c327544f2839e/corehq/apps/users/models.py#L1191-L1192
`clear_quickcache_for_user` and `delete_demo_restore_for_user` are also called on delete.

Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: Leaving this as is because I'd rather not copy in the cache clearing, etc. Will revisit this if this performs badly.

File: corehq/apps/users/views/mobile/users.py (hunk 6 above)
Comment (mkangia) [reply corehq/apps/users/views/mobile/users.py]: Sorry, functions :)

Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: broke this up into smaller functions in 36d3198235c

File: corehq/apps/users/views/mobile/users.py (hunk 7 above)
Comment (orangejenny) [reply corehq/apps/users/views/mobile/users.py]: Leaving this in the view, I think it's more common in HQ to process file contents in the view, not the form where it was uploaded (probably because we often mix validation & processing).

**Comment by orangejenny** [comment]
Thanks all for feedback, everything should be responded to now.

**Review by mkangia — APPROVED** [review]

**Review by snopoke — APPROVED** [review]
